Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE in wtf/text/WTFString.cpp

cdumez opened the pull request:
#### 4c227b3c248ed64abc1561bfc143c9172f5770a5
<pre>
Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE in wtf/text/WTFString.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=282909">https://bugs.webkit.org/show_bug.cgi?id=282909</a>

Reviewed by Darin Adler.

* Source/JavaScriptCore/assembler/LinkBuffer.cpp:
(JSC::LinkBuffer::finalizeCodeWithDisassemblyImpl):
* Source/JavaScriptCore/ftl/FTLAbstractHeap.cpp:
(JSC::FTL::IndexedAbstractHeap::initialize):
* Source/WTF/wtf/SHA1.cpp:
(WTF::SHA1::hexDigest):
* Source/WTF/wtf/SafeStrerror.cpp:
(WTF::safeStrerror):
* Source/WTF/wtf/persistence/PersistentCoders.cpp:
(WTF::Persistence::Coder&lt;CString&gt;::decodeForPersistence):
* Source/WTF/wtf/text/CString.cpp:
(WTF::CString::newUninitialized):
* Source/WTF/wtf/text/CString.h:
* Source/WTF/wtf/text/WTFString.cpp:
(WTF::String::ascii const):
(WTF::String::latin1 const):
(WTF::String::fromCodePoint):
* Source/WTF/wtf/win/FileSystemWin.cpp:
(WTF::FileSystemImpl::fileSystemRepresentation):
* Source/WebCore/PAL/pal/crypto/commoncrypto/CryptoDigestCommonCrypto.cpp:
(PAL::CryptoDigest::toHexString):
* Source/WebCore/PAL/pal/text/win/TextCodecWin.cpp:
(PAL::TextCodecWin::encode):
* Source/WebCore/platform/sql/SQLiteFileSystem.cpp:
(WebCore::SQLiteFileSystem::computeHashForFileName):
* Source/WebKit/Shared/mac/AuxiliaryProcessMac.mm:
(WebKit::tryApplyCachedSandbox):
* Tools/TestWebKitAPI/Tests/WTF/CString.cpp:
(TEST(WTF, CStringUninitializedConstructor)):

Canonical link: <a href="https://commits.webkit.org/286466@main">https://commits.webkit.org/286466@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f10ad433585c10875be533e056c696b2079cf1ed

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76057 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55086 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28956 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80557 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27323 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/78173 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64228 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3382 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59640 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17788 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79124 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49522 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65317 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39997 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46921 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22802 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25650 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/69234 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/68037 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23139 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82018 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/75333 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3426 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2198 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67869 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3580 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65285 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67180 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11125 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/9245 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/97587 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11771 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3376 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/6182 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21347 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3397 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/4936 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/4178 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/SHA1.cpp:
```diff
@@ -242,13 +242,10 @@ void SHA1::addUTF8Bytes(CFStringRef string)
 {
-    char* start = nullptr;
+    std::span<char> start;
     CString result = CString::newUninitialized(40, start);
```
I wonder why this uses the magic number 40. Seems safer to write `hashSize * 2`.

darinadler commented on Source/WTF/wtf/SHA1.cpp:
```diff
@@ -242,13 +242,10 @@ void SHA1::addUTF8Bytes(CFStringRef string)
-        buffer += 2;
-    }
+    for (size_t i = 0; i < hashSize; ++i)
+        snprintf(start.subspan(i * 2).data(), 3, "%02X", digest.at(i));
```
We have functions in `ASCIICType.h` that turn a byte into hex digits without involving the `snprintf` function; I think they should be used here instead. This code using `snprintf` is probably a bit slow, parsing that format string over and over again, and writes out null byte terminators that are not needed, mostly overwritten by the following hex digits and then the last one just overwrites a null character that is already put there by `CString::newUninitialized`.

darinadler commented on Source/JavaScriptCore/assembler/LinkBuffer.cpp:
```diff
@@ -145,13 +145,13 @@ LinkBuffer::CodeRef<LinkBufferPtrTag> LinkBuffer::finalizeCodeWithDisassemblyImp
-        snprintf(buffer, length, "%s", prefix);
-        vsnprintf(buffer + sizeof(prefix) - 1, stringLength + 1, format, argList);
-        out.printf("%s", buffer);
+        snprintf(buffer.data(), buffer.size(), "%s", prefix.characters());
```
This is a peculiar use of `snprintf` to just copy characters. I think `memcpySpan` would be better.

darinadler commented on Source/WTF/wtf/SafeStrerror.cpp:
```diff
@@ -36,23 +36,23 @@ namespace WTF {
         // We have GNU strerror_r(), which returns char*. This may or may not be a pointer into
         // cstringBuffer. We also have to be careful because this has to compile even if ret is
         // an int, hence the reinterpret_casts.
         char* message = reinterpret_cast<char*>(ret);
```
Just noticed the comment above about this having to compile if `ret` is `int` in passing and it seems wrong.

darinadler commented on Source/WebCore/PAL/pal/crypto/commoncrypto/CryptoDigestCommonCrypto.cpp:
```diff
@@ -221,14 +221,11 @@ String CryptoDigest::toHexString()
-        buffer += 2;
-    }
+    for (size_t i = 0; i < hashLength; ++i)
+        snprintf(start.subspan(i * 2).data(), 3, "%02X", hash.at(i));
```
This copied and pasted code should probably share a function with the one from `SHA1.cpp`. The same mistakes are here. But also here seems unnecessary to use a `CString` and turn it into a `String`. More efficient to make a Latin-1 `String` directly. So maybe need a separate `String` and `CString` version?

darinadler commented on Source/WebCore/platform/sql/SQLiteFileSystem.cpp:
```diff
@@ -157,14 +157,11 @@ String SQLiteFileSystem::computeHashForFileName(StringView fileName)
-        buffer += 2;
-    }
+    for (size_t i = 0; i < digestLength; ++i)
+        snprintf(start.subspan(i * 2).data(), 3, "%02X", digest.at(i));
```
This copied and pasted code should probably share a function with the one from `SHA1.cpp`. The same mistakes are here. But also here seems unnecessary to use a `CString` and turn it into a `String`. More efficient to make a Latin-1 `String` directly. So maybe need a separate `String` and `CString` version?

darinadler commented on Source/WTF/wtf/HexNumber.cpp:
```diff
@@ -46,13 +48,38 @@ std::pair<LChar*, unsigned> appendHex(LChar* buffer, unsigned bufferSize, std::u
 
-} // namespace WTF
+template<typename CharacterType>
+void toHexadecimalInternal(std::span<const uint8_t> values, std::span<CharacterType>& hexadecimalOutput)
```
Could use `byteCast` and make this a non-template function.

darinadler commented on Source/WTF/wtf/HexNumber.h:
```diff
@@ -87,12 +87,17 @@ template<> class StringTypeAdapter<HexNumberBuffer> {
 
 using WTF::hex;
+using WTF::toHexadecimalCString;
+using WTF::toHexadecimalString;
```
I think we usually just say "hex" rather than "hexadecimal".

darinadler commented on Source/WTF/wtf/HexNumber.cpp:
```diff
@@ -46,13 +49,37 @@ std::pair<LChar*, unsigned> appendHex(LChar* buffer, unsigned bufferSize, std::u
+CString toHexCString(std::span<const uint8_t> values)
+{
+    std::span<char> buffer;
+    auto result = CString::newUninitialized(values.size() * 2, buffer);
```
Just occurred to me that this could overflow in theory even though it’s really not a practical concern. Should it perhaps use checked arithmetic or have a `RELEASE_ASSERT`?

cdumez committed:
Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE in wtf/text/WTFString.cpp
https://bugs.webkit.org/show_bug.cgi?id=282909

Reviewed by Darin Adler.

* Source/JavaScriptCore/assembler/LinkBuffer.cpp:
(JSC::LinkBuffer::finalizeCodeWithDisassemblyImpl):
* Source/JavaScriptCore/ftl/FTLAbstractHeap.cpp:
(JSC::FTL::IndexedAbstractHeap::initialize):
* Source/WTF/wtf/SHA1.cpp:
(WTF::SHA1::hexDigest):
* Source/WTF/wtf/SafeStrerror.cpp:
(WTF::safeStrerror):
* Source/WTF/wtf/persistence/PersistentCoders.cpp:
(WTF::Persistence::Coder<CString>::decodeForPersistence):
* Source/WTF/wtf/text/CString.cpp:
(WTF::CString::newUninitialized):
* Source/WTF/wtf/text/CString.h:
* Source/WTF/wtf/text/WTFString.cpp:
(WTF::String::ascii const):
(WTF::String::latin1 const):
(WTF::String::fromCodePoint):
* Source/WTF/wtf/win/FileSystemWin.cpp:
(WTF::FileSystemImpl::fileSystemRepresentation):
* Source/WebCore/PAL/pal/crypto/commoncrypto/CryptoDigestCommonCrypto.cpp:
(PAL::CryptoDigest::toHexString):
* Source/WebCore/PAL/pal/text/win/TextCodecWin.cpp:
(PAL::TextCodecWin::encode):
* Source/WebCore/platform/sql/SQLiteFileSystem.cpp:
(WebCore::SQLiteFileSystem::computeHashForFileName):
* Source/WebKit/Shared/mac/AuxiliaryProcessMac.mm:
(WebKit::tryApplyCachedSandbox):
* Tools/TestWebKitAPI/Tests/WTF/CString.cpp:
(TEST(WTF, CStringUninitializedConstructor)):

Canonical link: https://commits.webkit.org/286466@main
